101-convolution.rst

```diff
@@ -0,0 +1,202 @@
+on the frontend side.
+It is essentially the addition of a new book-keeping object, with which we'd
+want to include some validation.
+But other than that, the addition of this new object shouldn't affect
+anything else on the frontend.
+
+Most of the effort will be on the backend.
+The reference builder would add a new operation (e.g. ``Conv2DInc``).
+The reference backend can implement this operation by expanding
+it to a dense weight matrix and then proceeding as normal with a ``DotInc``.
+We could make this fallback approach available for any backend that does not
+want to add a custom convolution implementation.
+We could perhaps add a more efficient implementation as well (e.g.,
+if the user has ``scipy`` installed we could use ``scipy.signal.convolve2d``).
+
+Most of the benefit will be for other backends that have specialized
+operations for convolution (e.g. NengoDL, NengoOCL, Nengo Loihi).
+These will be able to provide custom implementations for ``Conv2DInc`` (if
+they use the reference builder), or can treat the ``transform=nengo.Conv2D``
+object however they like if they are writing their own ``Connection`` builder.
+The implementation in these backends should proceed concurrently with this
+proposal, to be sure that the design here meets their needs.
+
+Open questions/extensions
+-------------------------
+
+**Passing these transform objects for other parameters**
+
+In the same way that we can pass Distributions for transforms or encoders,
+it seems natural to think about using these new objects for encoders as well
+(e.g., creating convolutional encoders).
+There doesn't seem to be anything particularly problematic about this, it would
+just be a matter of trying it out.
+
+**Using these objects on decoded connections**
+
+A little more tricky is how these objects should interact with decoders.
+Currently we combine the decoders and the transform on decoded connections into
+a single dense weight matrix.
+It wouldn't be possible to do that combination with a convolutional transform
+(short of expanding the convolutional weights into a dense weight matrix).
+One option would be to keep them distinct just in this special case.
+In this case we would have a somewhat significant qualitative change in the
+model when changing the transform type.
+Alternatively, we could separate the decoders and transform in all cases.
+This would affect online learning rules, as they would target one or the other
+rather than both (as they do now).
+This may also be less efficient for backends to implement, without additional
+optimization on their part.
+
+My inclination would be to leave the current behaviour unchanged and only
+separate the decoders/transform for convolutional transforms.
+But this issue seems more uncertain, and a resolution is not required in order
+to support the vast majority of convolutional use cases (which are based
+on neuron-to-neuron connections).
+That is why we have excluded this issue from this proposal.
```

Comment:
My inclination is to always combine the convolutional transform and decoders (using the weight matrix expansion), or just error in this case. I think there are enough other ways to get that same functionality, especially if we're allowing convolutional encoders (since that already splits the connection into two parts, decoders and encoders). I certainly don't think it's worth hitting our heads against the wall over, since I can't think of any use cases for a convolutional transform on a decoded connection off the top of my head (that couldn't be done with convolutional encoders).